src/Tools/dotnet-monitor/CollectionRules/CollectionRulePipeline.cs

```diff
@@ -0,0 +1,248 @@
+            _endpointInfo = endpointInfo ?? throw new ArgumentNullException(nameof(endpointInfo));
+            _ruleOptions = ruleOptions ?? throw new ArgumentNullException(nameof(ruleOptions));
+            _triggerOperations = triggerOperations ?? throw new ArgumentNullException(nameof(triggerOperations));
+        }
+
+        /// <summary>
+        /// Starts the execution of the pipeline without waiting for it to run to completion.
+        /// </summary>
+        /// <remarks>
+        /// If the specified trigger is a startup trigger, this method will complete when the
+        /// action list has completed execution. If the specified trigger is not a startup
+        /// trigger, this method will complete after the trigger has been started.
+        /// </remarks>
+        public async Task StartAsync(CancellationToken token)
+        {
+            var runTask = RunAsync(token);
+
+            await _startedSource.WithCancellation(token).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Runs the pipeline to completion.
+        /// </summary>
+        /// <remarks>
+        /// The pipeline will only successfully complete in the following scenarios:
+        /// (1) the trigger is a startup trigger and the action list successfully executes once.
+        /// (2) without a specified action count window duration, the number of action list executions equals the action count limit.
+        /// </remarks>
+        protected override async Task OnRun(CancellationToken token)
+        {
+            if (!_triggerOperations.TryCreateFactory(_ruleOptions.Trigger.Type, out ICollectionRuleTriggerFactoryProxy factory))
+            {
+                throw new InvalidOperationException(string.Format(CultureInfo.InvariantCulture, Strings.ErrorMessage_CouldNotMapToTrigger, _ruleOptions.Trigger.Type));
+            }
+
+            using CancellationTokenSource durationCancellationSource = new();
+            using CancellationTokenSource linkedCancellationSource = CancellationTokenSource.CreateLinkedTokenSource(
+                durationCancellationSource.Token,
+                token);
+
+            CancellationToken linkedToken = linkedCancellationSource.Token;
+
+            TimeSpan? actionCountWindowDuration = _ruleOptions.Limits?.ActionCountSlidingWindowDuration;
+            int actionCountLimit = (_ruleOptions.Limits?.ActionCount).GetValueOrDefault(CollectionRuleLimitsOptionsDefaults.ActionCount);
+            Queue<DateTime> executionTimestamps = new(actionCountLimit);
+
+            // Start cancellation timer for graceful stop of the collection rule
+            // when the rule duration has been specified. Conditionally enable this
+            // based on if the rule has a duration limit.
+            TimeSpan? ruleDuration = _ruleOptions.Limits?.RuleDuration;
+            if (ruleDuration.HasValue)
+            {
+                durationCancellationSource.CancelAfter(ruleDuration.Value);
```

Comment:
Shouldn't we only create this if there is a rule duration?

Reply:
Conditionally creating a linked token source that needs to be conditionally disposed requires some extra code and acrobatics to conditionally create it, set the cancellation time, and dispose it. The way I have it just creates the linked token source and only applies the cancellation due to duration if the rule has a specified duration. Maybe this can be cleaned up better with a refactor of the OnRun method into smaller components.